src/hotspot/share/opto/loopTransform.cpp

```diff
@@ -1595,6 +1683,166 @@ void PhaseIdealLoop::insert_vector_post_loop(IdealLoopTree *loop, Node_List &old
   // finds some, but we _know_ they are all useless.
   peeled_dom_test_elim(loop, old_new);
   loop->record_for_igvn();
+  C->print_method(PHASE_AFTER_ATOMIC_POST, 4, post_head);
+}
+
+//------------------------------insert_atomic_post_loop_impl-------------------------------
+// The main implementation of inserting atomic post loop after vector main loop.
+void PhaseIdealLoop::insert_atomic_post_loop_impl(IdealLoopTree* loop, Node_List& old_new,
+                                                  CountedLoopNode* main_head, CountedLoopEndNode* main_end,
+                                                  Node*& main_incr, Node* limit, CountedLoopNode*& post_head) {
+  IfNode* outer_main_end = main_end;
+  IdealLoopTree* outer_loop = loop;
+  if (main_head->is_strip_mined()) {
+    main_head->verify_strip_mined(1);
+    outer_main_end = main_head->outer_loop_end();
+    outer_loop = loop->_parent;
+    assert(outer_loop->_head == main_head->in(LoopNode::EntryControl), "broken loop tree");
+  }
+
+  Node* main_exit = outer_main_end->proj_out(false);
+  assert(main_exit->Opcode() == Op_IfFalse, "");
+  int dd_main_exit = dom_depth(main_exit);
+
+  // Step 1: Clone the loop body of vector main loop. The clone becomes the atomic post-loop.
+  const uint first_node_index_in_cloned_loop_body = C->unique();
+  clone_loop(loop, old_new, dd_main_exit, InsertAtomicPost);
+  assert(old_new[main_end->_idx]->Opcode() == Op_CountedLoopEnd, "");
+  post_head = old_new[main_head->_idx]->as_CountedLoop();
+  post_head->set_normal_loop();
+  post_head->set_post_loop(main_head);
+
+  // clone_loop() above changes the exit projection
+  main_exit = outer_main_end->proj_out(false);
```

Comment:
Looks like a lot of code duplication with `PhaseIdealLoop::insert_post_loop`. We maybe want to think how to refactor this.